test/helpers.js

```diff
@@ -175,6 +175,39 @@
     })
   });
 
+  MockServiceWithErrorCodeMapping = function(mapping) {
+    return AWS.Service.defineService('mock', {
+      serviceIdentifier: 'mock',
+      initialize: function (config) {
+        AWS.Service.prototype.initialize.call(this, config);
+        this.config.credentials = {
+          accessKeyId: 'akid',
+          secretAccessKey: 'secret'
+        };
+        return (this.config.region = 'mock-region');
+      },
+      setupRequestListeners: function (request) {
+        return request.on('extractError', function (resp) {
+          return (resp.error = {
+            code:
+              (resp.httpResponse.body || '').toString() ||
```

Comment:
This doesn't look like it matches runtime behaviour outside of tests. Would it make sense to have it grab the error from the JSON response payload?